File: c++/merge_sort.cpp
#include<iostream>
using namespace std;
void merge(int arr[], int left, int mid, int right) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    int* L = new int[n1];
    int* R = new int[n2];

    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    int i = 0;
    int j = 0;
    int k = left;

    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }

    delete[] L;
    delete[] R;
}
int main(){
    int arr[] = {38, 27, 43, 3, 9, 82, 10};
    int arr_size = sizeof(arr)/sizeof(arr[0]);

    cout << "Given array is \n";
    for(int i=0; i<arr_size; i++)
        cout << arr[i] << " ";
    cout << endl;

    merge(arr, 0, (arr_size - 1) / 2, arr_size - 1);

    cout << "\nSorted array is \n";
    for(int i=0; i<arr_size; i++)
        cout << arr[i] << " ";
    cout << endl;
    return 0;
}

File: go/merge_sort.go
package main

import (
    "fmt"
)

// merge combines two sorted slices into one sorted slice
func merge(left, right []int) []int {
    result := []int{}
    i, j := 0, 0

    // Compare elements from both slices and pick the smaller one
    for i < len(left) && j < len(right) {
        if left[i] < right[j] {
            result = append(result, left[i])
            i++
        } else {
            result = append(result, right[j])
            j++
        }
    }

    // Add any remaining elements
    result = append(result, left[i:]...)
    result = append(result, right[j:]...)

    return result
}

// mergeSort recursively splits and sorts the array
func mergeSort(arr []int) []int {
    if len(arr) <= 1 {
        return arr
    }

    mid := len(arr) / 2
    left := mergeSort(arr[:mid])
    right := mergeSort(arr[mid:])

    return merge(left, right)
}

func main() {
    var n int
    fmt.Print("Enter number of elements: ")
    fmt.Scan(&n)

    arr := make([]int, n)
    fmt.Println("Enter elements:")
    for i := 0; i < n; i++ {
        fmt.Scan(&arr[i])
    }

    sorted := mergeSort(arr)

    fmt.Println("Sorted array:")
    fmt.Println(sorted)
}
